ql/src/java/org/apache/hadoop/hive/ql/io/IOContext.java

```diff
@@ -51,7 +51,7 @@ public class IOContext {
    */
   private  RecordIdentifier ri;
   private boolean isDeletedRecord;
-  private PositionDeleteInfo pdi;
+  private ThreadLocal<PositionDeleteInfo> pdi = new ThreadLocal<>();
```

Comment:
shouldn't the whole `IOContext` be thread local or something?

Reply:
Should not be the whole IOContext class immutable? It contains many fields which can be modified asynchronously and it can cause concurrency issues.